### PR TITLE
feat: add support for Ibis 9.x 

### PR DESCRIPTION
**Pull request by tokoko**
- Adds support for Ibis 9.x and drops all previous versions.
- Adds `test_parity.py` checking execution parity between duckdb (with ibis), duckdb (with substrait) and acero.
- All but one TPCH queries are now green.

### CODE REVIEW

**Comment by tokoko** [comment]
Worked around arrow nightly failures by first using duckdb backend's `to_pandas` method and then converting to arrow. Looks like `to_pyarrow` doesn't work with nightly arrow. it's [passing](https://github.com/ibis-project/ibis/blob/main/ibis/formats/pyarrow.py#L259) a tuple to `rename_columns` method instead of a list. Should probably be "fixed" in ibis.

P.S. Still fighting nix about other failures...

**Comment by gforsyth** [comment]
Hey @tokoko -- I'm going to take a more thorough look at this soon, but just an FYI that the nix failures are due to an upstream thing that has been fixed but is still propagating down to the various nixos package repos, so we'll just have to wait another day or two for it to resolve itself.

**Comment by tokoko** [comment]
@gforsyth sure, thanks for the info. I think that's the reason for 39 and 310 failures. 311 failure is strange though, duckdb is failing to download substrait extension only when in nix environment, works without a problem in poetry. I'll also look into other failures as well. The windows one is not too surprising :) but I'm not sure what semantic-version is complaining about. Any ideas there?

**Comment by gforsyth** [comment]
> duckdb is failing to download substrait extension only when in nix environment, works without a problem in poetry

Ahh, that's a nix sandboxing thing -- we have test markers in Ibis (upstream) to skip tests that involve extensions when doing a nix build.




> I'm not sure what semantic-version is complaining about. Any ideas there?

Might just need to be updated in CI, let me check

**Comment by gforsyth** [comment]
> Might just need to be updated in CI, let me check

Ok, a bump to conventional commits sorted it out.

**Comment by gforsyth** [comment]
Hey @tokoko -- I bumped the nix flake and it pulled in the upstream fix, so the nix builds are passing now (with the added xfails of the tests that require installing the duckdb substrait extension on linux)

**Comment by tokoko** [comment]
@gforsyth I saw that, thanks. I'll try to later reorganize tests with fixtures in a way that only duckdb+substrait portion can be skipped instead of entire tests. I think I added a "fix" for windows failure as well.

**Comment by tokoko** [comment]
I tried but substrait extension doesn't seem to work on windows, I'm not even able to download the binary from the browser. Acero seems to support more functionality, it wasn't adding much value anyway, so I went ahead and disabled it from tests for now.

**Comment by tokoko** [comment]
@gforsyth Hey, have you had an opportunity to dig through the changes?

**Comment by gforsyth** [comment]
> @gforsyth Hey, have you had an opportunity to dig through the changes?

currently digging!  I have a few small changes to add, but nothing big.  I think the biggest pieces left are to get in tests for window functions and to figure out if we need to keep `_get_child_relation_field_offsets` for anything.  
That said, neither of those two things should block this PR.

File: ibis_substrait/compiler/translate.py
Comment (gforsyth) [inline ibis_substrait/compiler/translate.py:999]:
```suggestion
                measure=translate(agg_func, compiler=compiler, **kwargs)
```

File: ibis_substrait/compiler/translate.py
Comment (gforsyth) [inline ibis_substrait/compiler/translate.py:1001]:
```suggestion
            for agg_func in op.metrics.keys()
```

**Comment by tokoko** [comment]
@gforsyth Sweet, sounds good.

> get in tests for window functions

This is a bit of a pain as substrait consumer engines I used in parity tests didn't support them. Pretty sure datafusion supports window functions though, just couldn't get datafusion tests working yet.

> figure out if we need to keep `_get_child_relation_field_offsets` for anything

I sort of convinced myself that we don't, but more tests covering edge cases would be better.

File: ibis_substrait/tests/compiler/snapshots/test_tpch/test_compile/tpc_h07/tpc_h07.json
Comment (gforsyth) [inline ibis_substrait/tests/compiler/snapshots/test_tpch/test_compile/tpc_h07/tpc_h07.json:47]: these scalar functions disappearing is a bit odd -- I'll have to dig in deeper (in a follow-up) to make sure this is working correctly

**Review by gforsyth — COMMENTED** [review]
Really nice work here @tokoko -- thanks for tackling this!

Some cleanup comments, and a few questions for you, and a few questions for (future) me.

File: ibis_substrait/compiler/translate.py
Comment (tokoko) [reply ibis_substrait/compiler/translate.py:39]: Missed that. This is unrelated but are you going to cut straight to 9.1 dropping 3.9? Some of the dependent libraries (I care about feast of course 😄) will have trouble following so soon.

Comment (tokoko) [reply ibis_substrait/compiler/translate.py:39]: To be clear, at least a single release targeting ibis 9.0 with py3.9 support would be a lifesaver.

File: ibis_substrait/compiler/translate.py
Comment (tokoko) [reply ibis_substrait/compiler/translate.py]: This should be .values(), not .keys(). I'll fix in a follow-up

File: ibis_substrait/compiler/translate.py
Comment (tokoko) [reply ibis_substrait/compiler/translate.py]: Removed `ops.Sum` rule

File: ibis_substrait/compiler/translate.py
Comment (gforsyth) [reply ibis_substrait/compiler/translate.py:39]: we can definitely do that.

File: ibis_substrait/compiler/translate.py
Comment (gforsyth) [reply ibis_substrait/compiler/translate.py]: ahh, yeah, my mistake there

File: ibis_substrait/compiler/translate.py
Comment (tokoko) [reply ibis_substrait/compiler/translate.py:39]: great, thanks a lot

**Comment by tokoko** [comment]
Added handling for `countDistinct`, all tpch queries should be green now.

**Review by gforsyth — APPROVED** [review]
Thanks for all of your work on this, @tokoko !

I'm going to merge this once CI passes.  I'd like to add a few TPC-H parity tests just to make sure things are working as expected, but everything here looks good!

**Pull request merged**
